### DISCUSSION TITLE
locks/leases with KV similar to SETNX

### DISCUSSION THREAD

**Discussion by aep** [nats-io/nats-server#4803] [General]
with nats having KV now, i wonder if we can just drop redis entirely.
however, we use redis SETNX to build locks, (or leases in k8s terms)

the algo is well described here https://redis.io/docs/manual/patterns/distributed-locks/
and has been battle tested over the years.

has anyone done similar work with nats yet?
I guess we can somehow use Create() and that _should_ be atomic?
but there's no TTL, so its unclear how you would expire the lock if the locker is dead.

**Comment by jnmoyne** [comment]
You can do distributed synchronization and locked with NATS using the `DiscardNewPerSubject` discard policy with a limit of 1 message max per subject, and setting a max age for messages in the stream. 

The subject is the "key" that you want to lock, if you succeed to publish on the subject then you have acquired the lock, which you release by removing the message from the stream. If the publish fails then you could not get the lock. 

This is very similar to what is described in this blog: https://nats.io/blog/new-per-subject-discard-policy/ (basically a variation on the same theme). 

And yes you could also use the atomic compare and set features of JetStream (eg do an "insert" rather than "upsert") to the same effect.

**Reply by aep** [reply]
this is interesting, but doesnt appear to allow expiry?
once a lock is set, it is forever, so a dead client would require manual cleanup

**Reply by aep** [reply]
actually i guess i can combine this with MaxAge and hopefully a new send will reset the timer for this lock only.  going to experiment with it

**Reply by aep** [reply]
wow NATS is has so much stuff, i wonder if the whole thing can be done much smarter
i'm implementing a pattern similar to k8s "reconciliation" 

 - something is continuously pushing an absolute configuration into key A ( like color=green )
 - a worker occasionally looks at A whenever and picks the last value sent there, ignoring any intermediate message
 - it holds a lock to prevent other workers from working on the same topic A in the meantime
 - whether it succeeds or fails, the lock expires and another worker can pick whatever was pushed on A in the meantime

this seems to almost work with exactly-once-delivery, except that this appears to allow parallel delivery of _different_ messages to different consumers. It only prevents the same message to be replayed, not the same key.  I'm almost sure there's some existing pattern that allows me to implement this better

**Reply by ripienaar** [reply]
Look at the KV package it has Create() and Update() that's good for this - Create() to obtain the lock and then Update() to maintain it as long as needed.  Overall bucket TTL to expire stale locks.

Ofcourse while this approach is typically perfectly fine for majority of use cases it does not guarantee exclusivity but its on par with options in K8s etc that are widely used.

**Reply by aep** [reply]
instead of acquiring a lock, nats allows me to hand around a `token` that unlocks only one worker.

```
nats stream add  work --defaults --discard-per-subject --subjects='color.*' --storage=memory   --discard=new   --max-msgs-per-subject=1

nats req color.car red #ok
nats req color.car red #nope

nats con add work worker --ack=all --wait=5s --target="worker" --defaults --deliver-group=workers

nats sub worker --queue=workers
nats sub worker --queue=workers

# worker is done
nats stream rmm work 1

```

this is very nice, but the ack timeout is fixed. can i tell nats that to delay re-delivery for some more time while the worker is busy?

**Reply by ripienaar** [reply]
There is a Progress ack that you can use to tell it you're still busy - but this is in the libraries only not in CLI

**Reply by aep** [reply]
yes this works!

but is it correct? this is so simple, i feel like i must be not seeing something

```go
package main

import (
	"github.com/nats-io/nats.go"
	"time"
	"fmt"
)

func main() {

	nc, err := nats.Connect("localhost")
	if err != nil {
		panic(err)
	}

	js, err := nc.JetStream()
	if err != nil {
		panic(err)
	}

	js.AddStream(&nats.StreamConfig{
		Name:     "work",
		Subjects: []string{"color.*"},
		Storage:  nats.MemoryStorage,
		MaxMsgsPerSubject: 1,
		Discard: nats.DiscardNew,
	})

	ch := make(chan *nats.Msg, 64)
	nc.ChanQueueSubscribe("worker", "workers", ch)
	for msg := range ch {
		fmt.Printf("Received a message: %s\n", string(msg.Data))
		go func(msg *nats.Msg) {
			for i := 0; i < 30; i++ {
				fmt.Println("working...")
				msg.InProgress()
				time.Sleep(1 * time.Second)
			}
			msg.Ack()
			js.PurgeStream("work") 
		}(msg)
	}

}

```

**Reply by aep** [reply]
hmm one problem that show up immediately is that msg.InProgress() never fails.  its alwyas returning nil even if i kill the nats server. so this is not a reliable way to know if you're really holding a lock

**Reply by ripienaar** [reply]
Yeah, I think the server do respond though so should be possible to do something to get a confirm, need to check if thats the case though.

Try doing a `nc.Request(msg.Reply, []byte("+WPI"))` see if you get a response, thats what in progress does

**Reply by aep** [reply]
> nc.Request(msg.Reply, []byte("+WPI"))

this appears to reliably give me an error when the server is gone.
thanks so much, this looks very elegant.
going to poke around some more to see if there's any issues with this and wrap it up in a blog post.

**Comment by aep** [comment]
the [previous comment  ]( https://github.com/nats-io/nats-server/discussions/4803#discussioncomment-7612323)   describes the solution to building a lock with NATS. it is roughly:

```sh

nats stream add  work --defaults --discard-per-subject --subjects='locks.*' --storage=memory   --discard=new   --max-msgs-per-subject=1

nats req locks.bob red # ok
nats req locks.bob red # will fail

nats stream rmm locks 1 #clear the lock

```

however, with the help of the maintainers i discovered a much smarter method of ensuring only one worker runs on one item 
 exclusively

writeup  with more context is on my blog https://docs.kraudcloud.com/blog/2023/11/22/exclusive-worker-tokens-with-nats/


the basic idea is that the same DiscardNewPerSubject construct can be used to push a notification **and** the lock to the workers. nats will hold the item in the queue until the worker acks it, leading to an exclusive lock on the item.


```go
package main

import (
    "fmt"
    "github.com/nats-io/nats.go"
    "time"
)

func InitNats() {
    nc, err := nats.Connect("localhost")
    if err != nil {
        panic(err)
    }
    defer nc.Close()

    js, err := nc.JetStream()
    if err != nil {
        panic(err)
    }

    // first create an inbox queue holding the latest state of work to be done
    // values in here are replaced when new work on the same topic is submitted
    _, err = js.AddStream(&nats.StreamConfig{
        Name:     "inbox",
        Subjects: []string{"inbox.*"},

        MaxMsgsPerSubject: 1,
        Discard:           nats.DiscardNew,
    })
    if err != nil {
        panic(fmt.Sprintf("Error creating jetstream [needs a nats-server with -js] : %v", err))
    }

    // items are moved from the inbox into a token lock.
    // these are held by a worker until its done and only THEN a new value is pulled from the inbox.
    // if the worker fails to ack the item, it is resent to a different worker
    _, err = js.AddStream(&nats.StreamConfig{
        Name: "work",
        Sources: []*nats.StreamSource{
            {
                Name: "inbox",
            },
        },

        MaxMsgsPerSubject: 1,
        Discard:           nats.DiscardNew,

        // this means you cant update a running token
        DiscardNewPerSubject: true,

        // an ack deletes the message and frees the topic for new work
        Retention: nats.WorkQueuePolicy,
    })
    if err != nil {
        panic(fmt.Sprintf("Error creating jetstream [needs a nats-server with -js] : %v", err))
    }

    // push the token into a delivery group
    _, err = js.AddConsumer("work", &nats.ConsumerConfig{
        Durable:        "work",
        DeliverSubject: "work",
        DeliverGroup:   "workers",
        DeliverPolicy:  nats.DeliverAllPolicy,
        AckPolicy:      nats.AckExplicitPolicy,
        AckWait:        30 * time.Second,
        Heartbeat:      time.Second,
    })
    if err != nil {
        panic(fmt.Sprintf("Error creating jetstream consumer : %v", err))
    }

    ch := make(chan *nats.Msg, 64)
    nc.ChanQueueSubscribe("work", "workers", ch)

    for msg := range ch {

        if len(msg.Reply) == 0 {
            // not jetstream, probably keepalive
            continue
        }

        fmt.Println(msg.Reply)

        fmt.Printf("Received a message on %s: %s\n", msg.Subject, string(msg.Data))
        go func(msg *nats.Msg) {
            for i := 0; i < 60; i++ {
                fmt.Println("working...")
                rsp, err := nc.Request(msg.Reply, []byte("+WPI"), time.Second)
                if err != nil {
                    // lost lock, stop immediately or we risk working in parallel
                    panic(err)
                }
                fmt.Println("got in progress response", string(rsp.Data))
                time.Sleep(1 * time.Second)
            }
            fmt.Println("done")
            msg.Ack()
        }(msg)
    }
}
```

**Comment by pySilver** [comment]
@aep have you tried this in production?

**Reply by aep** [reply]
Unfortunately no. We're just using temporal now. 
However, I tested this alot, and I didn't discover any issues with the idea.

**Reply by rafsawicki** [reply]
FYI (and other people coming here from google), I've tested this setup on our platform and ended up having to discard it due to several issues (I did keep the inbox stream, but used it alongside distributed locks). It somewhat works only if messages are processed fast in the token queue, but it has issues if processing takes longer than the time for the next message to arrive in inbox. When a new message is published to inbox, NATS will attempt to publish message to a token stream, which fails if a token still being processed. A warning will be printed in logs in this case, and since publish is retried every 2 seconds, this results in a huge spam in logs.

```
[6] 2025/03/19 13:14:42.122826 [WRN] Error processing inbound source "inbox > >" for '$G' > 'token': maximum messages per subject exceeded
[6] 2025/03/19 13:14:44.290533 [WRN] Error processing inbound source "inbox > >" for '$G' > 'token': maximum messages per subject exceeded
[6] 2025/03/19 13:14:46.477033 [WRN] Error processing inbound source "inbox > >" for '$G' > 'token': maximum messages per subject exceeded
```
Besides warning messages in server logs, I believe that this can also cause the publishing process to stop completely if there are too many pending messages.

I've tried manually publishing messages from the code to token stream, which worked ok-ish, until I ran out of configured retries for some messages. Despite token stream using work queue policy, when message was not processed by worker due to running out of configured retires, the message stayed on the queue and wasn't removed, blocking further processing for that subject. Perhaps this is fixed now, but this made me discard the pattern and use another stream as a distributed lock (it has to be in memory though, due to performance issues).
